pkg/handler/common/binding.go

```diff
@@ -0,0 +1,220 @@
+		}
+		if roleUser.Group != "" && subject.Name == roleUser.Group {
+			return nil, errors.NewBadRequest("group %s already connected to role %s", roleUser.Group, roleID)
+		}
+	}
+
+	if roleUser.UserEmail != "" {
+		existingRoleBinding.Subjects = append(existingRoleBinding.Subjects,
+			rbacv1.Subject{
+				Kind:     rbacv1.UserKind,
+				APIGroup: rbacv1.GroupName,
+				Name:     roleUser.UserEmail,
+			})
+	}
+	if roleUser.Group != "" {
+		existingRoleBinding.Subjects = append(existingRoleBinding.Subjects,
+			rbacv1.Subject{
+				Kind:     rbacv1.GroupKind,
+				APIGroup: rbacv1.GroupName,
+				Name:     roleUser.Group,
+			})
+	}
+
+	if err := client.Patch(ctx, existingRoleBinding, ctrlruntimeclient.MergeFrom(oldBinding)); err != nil {
+		return nil, fmt.Errorf("failed to update role binding: %v", err)
+	}
+
+	return convertInternalRoleBindingToExternal(existingRoleBinding), nil
+}
+
+func BindUserToClusterRoleEndpoint(ctx context.Context, userInfoGetter provider.UserInfoGetter, projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider, clusterRoleUser apiv1.ClusterRoleUser, projectID, clusterID, roleID string) (interface{}, error) {
+	clusterProvider := ctx.Value(middleware.ClusterProviderContextKey).(provider.ClusterProvider)
+
+	cluster, err := GetCluster(ctx, projectProvider, privilegedProjectProvider, userInfoGetter, projectID, clusterID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := common.GetClusterClient(ctx, userInfoGetter, clusterProvider, cluster, projectID)
+	if err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Name: roleID}, &rbacv1.ClusterRole{}); err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	clusterRoleBindingList := &rbacv1.ClusterRoleBindingList{}
+	if err := client.List(ctx, clusterRoleBindingList, ctrlruntimeclient.MatchingLabels{UserClusterComponentKey: UserClusterBindingComponentValue}); err != nil {
+		return nil, common.KubernetesErrorToHTTPError(err)
+	}
+
+	var existingClusterRoleBinding *rbacv1.ClusterRoleBinding
+	for _, clusterRoleBinding := range clusterRoleBindingList.Items {
+		if clusterRoleBinding.RoleRef.Name == roleID {
+			existingClusterRoleBinding = clusterRoleBinding.DeepCopy()
+			break
+		}
+	}
+
+	if existingClusterRoleBinding == nil {
+		return nil, fmt.Errorf("the cluster role binding not found")
+	}
+
+	oldBinding := existingClusterRoleBinding.DeepCopy()
+	for _, subject := range existingClusterRoleBinding.Subjects {
+		if clusterRoleUser.UserEmail != "" && subject.Name == clusterRoleUser.UserEmail {
+			return nil, errors.NewBadRequest("user %s already connected to role %s", clusterRoleUser.UserEmail, roleID)
+		}
+		if clusterRoleUser.Group != "" && subject.Name == clusterRoleUser.Group {
+			return nil, errors.NewBadRequest("group %s already connected to role %s", clusterRoleUser.Group, roleID)
```

Comment:
return nil, errors.NewBadRequest("group %s already connected to cluster role %s", clusterRoleUser.Group, roleID)

Reply:
done